File: a2a_agents/python/a2ui_agent/src/a2ui/extension/validation.py
from typing import Any, Dict, List, Set, Union
import jsonschema

def validate_a2ui_json(a2ui_json: Union[Dict[str, Any], List[Any]], a2ui_schema: Dict[str, Any]) -> None:
    """
    Validates the A2UI JSON payload against the provided schema and checks for integrity.

    Args:
        a2ui_json: The JSON payload to validate.
        a2ui_schema: The schema to validate against.

    Raises:
        jsonschema.ValidationError: If the payload does not match the schema.
        ValueError: If integrity or topology checks fail.
    """
    jsonschema.validate(instance=a2ui_json, schema=a2ui_schema)

    # Normalize to list for iteration
    messages = a2ui_json if isinstance(a2ui_json, list) else [a2ui_json]

    for message in messages:
        if not isinstance(message, dict):
            continue
            
        # Check for SurfaceUpdate which has 'components'
        if "components" in message:
            _validate_component_integrity(message["components"], a2ui_schema)
            _validate_topology(message["components"], a2ui_schema)
            
        _validate_recursion_and_paths(message)


def _validate_component_integrity(components: List[Dict[str, Any]], a2ui_schema: Dict[str, Any] = None) -> None:
    """
    Validates that:
    1. All component IDs are unique.
    2. A 'root' component exists.
    3. All references (children, child, etc.) point to existing IDs.
    """
    ids: Set[str] = set()
    
    # 1. Collect IDs and check for duplicates
    for comp in components:
        comp_id = comp.get("id")
        if comp_id is None:
            continue
            
        if comp_id in ids:
            raise ValueError(f"Duplicate component ID found: '{comp_id}'")
        ids.add(comp_id)

    # 2. Check for root component
    if "root" not in ids:
         raise ValueError("Missing 'root' component: One component must have 'id' set to 'root'.")

    # 3. Check for dangling references using schema-driven extraction
    ref_fields_map = _extract_component_ref_fields(a2ui_schema) if a2ui_schema else {}
    
    for comp in components:
        comp_props_container = comp.get("componentProperties")
        if not isinstance(comp_props_container, dict):
            continue
            
        for comp_type, props in comp_props_container.items():
            if not isinstance(props, dict):
                continue

            # Determine fields to check for this component type
            # Strictly use schema; if not found, assume no references (generic schema support)
            single_refs, list_refs = ref_fields_map.get(comp_type, (set(), set()))
            
            for key, value in props.items():
                if key in single_refs:
                    if isinstance(value, str) and value not in ids:
                         raise ValueError(f"Component '{comp.get('id')}' references missing ID '{value}' in field '{key}'")
                elif key in list_refs:
                    if isinstance(value, list):
                        for item in value:
                             if isinstance(item, str) and item not in ids:
                                 raise ValueError(f"Component '{comp.get('id')}' references missing ID '{item}' in field '{key}'")


def _validate_topology(components: List[Dict[str, Any]], a2ui_schema: Dict[str, Any] = None) -> None:
    """
    Validates the topology of the component tree:
    1. No circular references (including self-references).
    2. No orphaned components (all components must be reachable from 'root').
    """
    adj_list: Dict[str, List[str]] = {}
    all_ids: Set[str] = set()
    
    ref_fields_map = _extract_component_ref_fields(a2ui_schema) if a2ui_schema else {}

    # Build Adjacency List
    for comp in components:
        comp_id = comp.get("id")
        if comp_id is None:
            continue
        
        all_ids.add(comp_id)
        if comp_id not in adj_list:
            adj_list[comp_id] = []
        
        comp_props_container = comp.get("componentProperties")
        if not isinstance(comp_props_container, dict):
            continue
            
        for comp_type, props in comp_props_container.items():
            if not isinstance(props, dict):
                continue
            
            # Determine fields to check
            # Strictly use schema
            single_refs, list_refs = ref_fields_map.get(comp_type, (set(), set()))

            for key, value in props.items():
                if key in single_refs:
                    if isinstance(value, str):
                        if value == comp_id:
                            raise ValueError(f"Self-reference detected: Component '{comp_id}' references itself in field '{key}'")
                        adj_list[comp_id].append(value)
                elif key in list_refs:
                    if isinstance(value, list):
                        for item in value:
                            if isinstance(item, str):
                                if item == comp_id:
                                    raise ValueError(f"Self-reference detected: Component '{comp_id}' references itself in field '{key}'")
                                adj_list[comp_id].append(item)

    # Detect Cycles using DFS
    visited: Set[str] = set()
    recursion_stack: Set[str] = set()

    def dfs(node_id: str):
        visited.add(node_id)
        recursion_stack.add(node_id)
        
        for neighbor in adj_list.get(node_id, []):
            if neighbor not in visited:
                dfs(neighbor)
            elif neighbor in recursion_stack:
                raise ValueError(f"Circular reference detected involving component '{neighbor}'")
        
        recursion_stack.remove(node_id)

    if "root" in all_ids:
        dfs("root")

    # Check for Orphans
    orphans = all_ids - visited
    if orphans:
        sorted_orphans = sorted(list(orphans))
        raise ValueError(f"Orphaned components detected (not reachable from 'root'): {sorted_orphans}")


def _extract_component_ref_fields(schema: Dict[str, Any]) -> Dict[str, tuple[Set[str], Set[str]]]:
    """
    Parses the JSON schema to identify which component properties reference other components.
    Returns a map: { component_name: (set_of_single_ref_fields, set_of_list_ref_fields) }
    """
    # print(f"DEBUG: _extract_component_ref_fields called with schema keys: {schema.keys()}")
    ref_map = {}
    
    # We expect schema structure to have 'properties' -> 'components' -> 'items' -> ...
    # OR we might be passed the root schema.
    # A typical A2UI schema has definitions in $defs or definitions.
    
    # 1. Locate component definitions
    # In the testing schema (and likely real one), component properties are inside:
    # properties -> components -> items -> properties -> componentProperties -> properties -> [ComponentType]
    
    try:
        # Navigate to componentProperties definitions in the schema
        # This path depends on the exact schema structure provided to validate_a2ui_json
        # We'll try to walk down standard paths.
        
        root_defs = schema.get("$defs") or schema.get("definitions", {})
        
        # Helper to check if a property schema looks like a ComponentId reference
        def is_component_id_ref(prop_schema: Dict[str, Any]) -> bool:
            ref = prop_schema.get("$ref", "")
            if ref.endswith("ComponentId"):
                return True
            # Check if it's an expanded schema that refs ComponentId (unlikely for direct prop but possible)
            return False

        def is_child_list_ref(prop_schema: Dict[str, Any]) -> bool:
            ref = prop_schema.get("$ref", "")
            if ref.endswith("ChildList"):
                return True
            # Or array of ComponentIds
            if prop_schema.get("type") == "array":
                items = prop_schema.get("items", {})
                if is_component_id_ref(items):
                    return True
            return False

        # Find where components are defined. 
        # In the provided common_types.json / standard_catalog.json, components are usually in a specific location
        # but the schema passed to validate_a2ui_json *is* the message schema which *includes* catalog definitions 
        # often via $defs or inline? 
        # Actually message schema has 'components' array items matching catalog entries.
        
        # Let's search for 'componentProperties' in the schema
        # We can implement a search or targeted lookup
        
        # Target path: properties.components.items.properties.componentProperties.properties
        # (This matches the structure in test_validation.py)
        
        comps_schema = schema.get("properties", {}).get("components", {})
        items_schema = comps_schema.get("items", {})
        comp_props_schema = items_schema.get("properties", {}).get("componentProperties", {})
        all_components = comp_props_schema.get("properties", {})
        
        for comp_name, comp_schema in all_components.items():
            single_refs = set()
            list_refs = set()
            
            props = comp_schema.get("properties", {})
            for prop_name, prop_schema in props.items():
                if is_component_id_ref(prop_schema):
                    single_refs.add(prop_name)
                elif is_child_list_ref(prop_schema):
                    list_refs.add(prop_name)
            
            if single_refs or list_refs:
                ref_map[comp_name] = (single_refs, list_refs)
    except Exception:
        # If schema traversal fails (structure mismatch), return empty to trigger fallback
        pass
        
    return ref_map


def _validate_recursion_and_paths(data: Any) -> None:
    """
    Validates:
    1. Global recursion depth limit (50).
    2. FunctionCall recursion depth limit (5).
    3. Path syntax for DataBindings/DataModelUpdates.
    """
    def traverse(item: Any, global_depth: int, func_depth: int):
        if global_depth > 50:
            raise ValueError("Global recursion limit exceeded: Depth > 50")

        if isinstance(item, list):
            for x in item:
                traverse(x, global_depth + 1, func_depth)
        elif isinstance(item, dict):
            # Check for path
            if "path" in item and isinstance(item["path"], str):
                 _validate_path_syntax(item["path"])
            
            # Check for FunctionCall (heuristic: has 'call' and 'args')
            is_func = "call" in item and "args" in item
            
            if is_func:
                if func_depth >= 5:
                    raise ValueError("Recursion limit exceeded: FunctionCall depth > 5")
                
                # Increment func_depth only for this branch, but global_depth matches traversal
                for k, v in item.items():
                    if k == "args":
                        traverse(v, global_depth + 1, func_depth + 1)
                    else:
                        traverse(v, global_depth + 1, func_depth)
            else:
                for v in item.values():
                    traverse(v, global_depth + 1, func_depth)

    traverse(data, 0, 0)


def _validate_path_syntax(path: str) -> None:
    """
    Validates that the path is either a valid JSON Pointer (starts with /)
    or a valid relative path (no empty segments).
    Also checks for spaces.
    """
    if not path:
        return
    
    parts = path.split('/')
    if path.startswith('/'):
        # JSON Pointer - starts with /
        # We allow empty keys (//) as per RFC 6901 but typical A2UI usage might not.
        # For now, we only enforce that it starts with / if intended to be absolute.
        pass
    else:
        # Relative path - should not have empty segments
        if any(not p for p in parts):
             raise ValueError(f"Invalid data model path: '{path}' contains empty segments (//)")
    
    if " " in path:
         raise ValueError(f"Invalid data model path: '{path}' contains spaces")

File: a2a_agents/java/src/main/java/org/a2ui/Validation.java
package org.a2ui;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion;
import com.networknt.schema.ValidationMessage;

import java.util.*;
import java.util.regex.Pattern;

/**
 * Validates A2UI JSON payloads against the provided schema and checks for integrity.
 */
public final class Validation {

    // RFC 6901 compliant regex for JSON Pointer
    private static final Pattern JSON_POINTER_PATTERN = Pattern.compile("^(?:/(?:[^~/]|~[01])*)*$");

    // Recursion Limits
    private static final int MAX_GLOBAL_DEPTH = 50;
    private static final int MAX_FUNC_CALL_DEPTH = 5;

    // Constants
    private static final String COMPONENTS = "components";
    private static final String ID = "id";
    private static final String COMPONENT_PROPERTIES = "componentProperties";
    private static final String ROOT = "root";
    private static final String PATH = "path";
    private static final String FUNCTION_CALL = "functionCall";
    private static final String CALL = "call";
    private static final String ARGS = "args";

    private static final ObjectMapper mapper = new ObjectMapper();

    private Validation() {}

    /**
     * Validates the A2UI JSON payload against the provided schema and checks for integrity.
     *
     * @param a2uiJson The JSON payload to validate (List or Map).
     * @param a2uiSchema The schema object to validate against.
     * @throws IllegalArgumentException if validation fails.
     */
    public static void validateA2uiJson(Object a2uiJson, Map<String, Object> a2uiSchema) {
        // 1. JSON Schema Validation using networknt
        JsonSchemaFactory factory = JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.V202012);
        JsonNode schemaNode = mapper.valueToTree(a2uiSchema);
        JsonSchema schema = factory.getSchema(schemaNode);

        JsonNode instanceNode = mapper.valueToTree(a2uiJson);
        Set<ValidationMessage> errors = schema.validate(instanceNode);
        if (!errors.isEmpty()) {
            StringBuilder sb = new StringBuilder("JSON Schema Validation Failed:\n");
            for (ValidationMessage error : errors) {
                sb.append("- ").append(error.getMessage()).append("\n");
            }
            throw new IllegalArgumentException(sb.toString().trim());
        }

        // Normalize to list for iteration
        List<Object> messages = new ArrayList<>();
        if (a2uiJson instanceof List) {
            messages.addAll((List<?>) a2uiJson);
        } else {
            messages.add(a2uiJson);
        }

        for (Object msgItem : messages) {
            if (!(msgItem instanceof Map)) {
                continue;
            }
            @SuppressWarnings("unchecked")
            Map<String, Object> message = (Map<String, Object>) msgItem;

            // Check for SurfaceUpdate which has 'components'
            if (message.containsKey(COMPONENTS)) {
                Object compsObj = message.get(COMPONENTS);
                if (compsObj instanceof List) {
                    @SuppressWarnings("unchecked")
                    List<Map<String, Object>> components = (List<Map<String, Object>>) compsObj;
                    Map<String, RefFields> refMap = extractComponentRefFields(a2uiSchema);
                    validateComponentIntegrity(components, refMap);
                    validateTopology(components, refMap);
                }
            }
            validateRecursionAndPaths(message);
        }
    }

    private static void validateComponentIntegrity(List<Map<String, Object>> components, Map<String, RefFields> refFieldsMap) {
        Set<String> ids = new HashSet<>();

        // 1. Collect IDs and check for duplicates
        for (Map<String, Object> comp : components) {
            Object compIdObj = comp.get(ID);
            if (!(compIdObj instanceof String)) continue;
            String compId = (String) compIdObj;

            if (ids.contains(compId)) {
                throw new IllegalArgumentException("Duplicate component ID found: '" + compId + "'");
            }
            ids.add(compId);
        }

        // 2. Check for root component
        if (!ids.contains(ROOT)) {
            throw new IllegalArgumentException("Missing '" + ROOT + "' component: One component must have '" + ID + "' set to '" + ROOT + "'.");
        }

        // 3. Check for dangling references using helper
        for (Map<String, Object> comp : components) {
            for (Map.Entry<String, String> ref : getComponentReferences(comp, refFieldsMap)) {
                String refId = ref.getKey();
                String fieldName = ref.getValue();
                if (!ids.contains(refId)) {
                    throw new IllegalArgumentException("Component '" + comp.get(ID) + "' references missing ID '" + refId + "' in field '" + fieldName + "'");
                }
            }
        }
    }

    private static void validateTopology(List<Map<String, Object>> components, Map<String, RefFields> refFieldsMap) {
        Map<String, List<String>> adjList = new HashMap<>();
        Set<String> allIds = new HashSet<>();

        // Build Adjacency List
        for (Map<String, Object> comp : components) {
            Object compIdObj = comp.get(ID);
            if (!(compIdObj instanceof String)) continue;
            String compId = (String) compIdObj;

            allIds.add(compId);
            adjList.putIfAbsent(compId, new ArrayList<>());

            for (Map.Entry<String, String> ref : getComponentReferences(comp, refFieldsMap)) {
                String refId = ref.getKey();
                String fieldName = ref.getValue();
                if (refId.equals(compId)) {
                    throw new IllegalArgumentException("Self-reference detected: Component '" + compId + "' references itself in field '" + fieldName + "'");
                }
                adjList.get(compId).add(refId);
            }
        }

        // Detect Cycles using DFS
        Set<String> visited = new HashSet<>();
        Set<String> recursionStack = new HashSet<>();

        if (allIds.contains(ROOT)) {
            dfs(ROOT, adjList, visited, recursionStack);
        }

        // Check for Orphans
        Set<String> orphans = new HashSet<>(allIds);
        orphans.removeAll(visited);
        if (!orphans.isEmpty()) {
            List<String> sortedOrphans = new ArrayList<>(orphans);
            Collections.sort(sortedOrphans);
            throw new IllegalArgumentException("Orphaned components detected (not reachable from '" + ROOT + "'): " + sortedOrphans);
        }
    }

    private static void dfs(String nodeId, Map<String, List<String>> adjList, Set<String> visited, Set<String> recursionStack) {
        visited.add(nodeId);
        recursionStack.add(nodeId);

        List<String> neighbors = adjList.getOrDefault(nodeId, Collections.emptyList());
        for (String neighbor : neighbors) {
            if (!visited.contains(neighbor)) {
                dfs(neighbor, adjList, visited, recursionStack);
            } else if (recursionStack.contains(neighbor)) {
                throw new IllegalArgumentException("Circular reference detected involving component '" + neighbor + "'");
            }
        }

        recursionStack.remove(nodeId);
    }

    private static class RefFields {
        Set<String> singleRefs = new HashSet<>();
        Set<String> listRefs = new HashSet<>();
    }

    @SuppressWarnings("unchecked")
    private static Map<String, RefFields> extractComponentRefFields(Map<String, Object> schema) {
        Map<String, RefFields> refMap = new HashMap<>();

        Map<String, Object> compsSchema = getMap(schema, "properties", COMPONENTS);
        if (compsSchema == null) return refMap;

        Map<String, Object> itemsSchema = getMap(compsSchema, "items");
        if (itemsSchema == null) return refMap;

        Map<String, Object> compPropsSchema = getMap(itemsSchema, "properties", COMPONENT_PROPERTIES);
        if (compPropsSchema == null) return refMap;

        Map<String, Object> allComponents = getMap(compPropsSchema, "properties");
        if (allComponents == null) return refMap;

        for (Map.Entry<String, Object> compEntry : allComponents.entrySet()) {
            String compName = compEntry.getKey();
            if (!(compEntry.getValue() instanceof Map)) continue;
            Map<String, Object> compSchema = (Map<String, Object>) compEntry.getValue();

            RefFields refs = new RefFields();
            Map<String, Object> props = getMap(compSchema, "properties");
            if (props != null) {
                for (Map.Entry<String, Object> propEntry : props.entrySet()) {
                    String propName = propEntry.getKey();
                    if (!(propEntry.getValue() instanceof Map)) continue;
                    Map<String, Object> propSchema = (Map<String, Object>) propEntry.getValue();

                    if (isComponentIdRef(propSchema)) {
                        refs.singleRefs.add(propName);
                    } else if (isChildListRef(propSchema)) {
                        refs.listRefs.add(propName);
                    }
                }
            }
            if (!refs.singleRefs.isEmpty() || !refs.listRefs.isEmpty()) {
                refMap.put(compName, refs);
            }
        }
        return refMap;
    }

    private static boolean isComponentIdRef(Map<String, Object> propSchema) {
        Object ref = propSchema.get("$ref");
        return ref instanceof String && ((String) ref).endsWith("ComponentId");
    }

    @SuppressWarnings("unchecked")
    private static boolean isChildListRef(Map<String, Object> propSchema) {
        Object ref = propSchema.get("$ref");
        if (ref instanceof String && ((String) ref).endsWith("ChildList")) {
            return true;
        }
        if ("array".equals(propSchema.get("type"))) {
            Map<String, Object> items = getMap(propSchema, "items");
            if (items != null && isComponentIdRef(items)) {
                return true;
            }
        }
        return false;
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> getMap(Map<String, Object> map, String... keys) {
        Map<String, Object> current = map;
        for (String key : keys) {
            Object val = current.get(key);
            if (val instanceof Map) {
                current = (Map<String, Object>) val;
            } else {
                return null;
            }
        }
        return current;
    }

    @SuppressWarnings("unchecked")
    private static List<Map.Entry<String, String>> getComponentReferences(Map<String, Object> component, Map<String, RefFields> refFieldsMap) {
        List<Map.Entry<String, String>> references = new ArrayList<>();
        Object compPropsContainerObj = component.get(COMPONENT_PROPERTIES);
        if (!(compPropsContainerObj instanceof Map)) {
            return references;
        }

        Map<String, Object> compPropsContainer = (Map<String, Object>) compPropsContainerObj;

        for (Map.Entry<String, Object> typeEntry : compPropsContainer.entrySet()) {
            String compType = typeEntry.getKey();
            if (!(typeEntry.getValue() instanceof Map)) continue;
            Map<String, Object> props = (Map<String, Object>) typeEntry.getValue();

            RefFields refs = refFieldsMap.getOrDefault(compType, new RefFields());

            for (Map.Entry<String, Object> propEntry : props.entrySet()) {
                String key = propEntry.getKey();
                Object value = propEntry.getValue();

                if (refs.singleRefs.contains(key) && value instanceof String) {
                    references.add(new AbstractMap.SimpleEntry<>((String) value, key));
                } else if (refs.listRefs.contains(key) && value instanceof List) {
                    for (Object item : (List<?>) value) {
                        if (item instanceof String) {
                            references.add(new AbstractMap.SimpleEntry<>((String) item, key));
                        }
                    }
                }
            }
        }
        return references;
    }

    private static void validateRecursionAndPaths(Object data) {
        traverse(data, 0, 0);
    }

    @SuppressWarnings("unchecked")
    private static void traverse(Object item, int globalDepth, int funcDepth) {
        if (globalDepth > MAX_GLOBAL_DEPTH) {
            throw new IllegalArgumentException("Global recursion limit exceeded: Depth > " + MAX_GLOBAL_DEPTH);
        }

        if (item instanceof List) {
            for (Object x : (List<?>) item) {
                traverse(x, globalDepth + 1, funcDepth);
            }
            return;
        }

        if (item instanceof Map) {
            Map<String, Object> map = (Map<String, Object>) item;

            // Check for path
            if (map.containsKey(PATH) && map.get(PATH) instanceof String) {
                String path = (String) map.get(PATH);
                if (!JSON_POINTER_PATTERN.matcher(path).matches()) {
                    throw new IllegalArgumentException("Invalid JSON Pointer syntax: '" + path + "'");
                }
            }

            // Check for FunctionCall
            boolean isFunc = map.containsKey(CALL) && map.containsKey(ARGS);

            if (isFunc) {
                if (funcDepth >= MAX_FUNC_CALL_DEPTH) {
                    throw new IllegalArgumentException("Recursion limit exceeded: " + FUNCTION_CALL + " depth > " + MAX_FUNC_CALL_DEPTH);
                }

                for (Map.Entry<String, Object> entry : map.entrySet()) {
                    String k = entry.getKey();
                    Object v = entry.getValue();
                    if (ARGS.equals(k)) {
                        traverse(v, globalDepth + 1, funcDepth + 1);
                    } else {
                        traverse(v, globalDepth + 1, funcDepth);
                    }
                }
            } else {
                for (Object v : map.values()) {
                    traverse(v, globalDepth + 1, funcDepth);
                }
            }
        }
    }
}
